src/EventHandler.ts

```diff
@@ -0,0 +1,248 @@
+    public onMiddleTap: (event: UIEvent) => void = () => {};
+    public onRightTap: (event: UIEvent) => void = () => {};
+    public onBackwardSwipe: (event: UIEvent) => void = () => {};
+    public onForwardSwipe: (event: UIEvent) => void = () => {};
+
+    public onLeftHover: () => void = () => {};
+    public onRightHover: () => void = () => {};
+    public onRemoveHover: () => void = () => {};
+
+    public constructor() {
+        this.pendingMouseEventStart = null;
+        this.pendingMouseEventEnd = null;
+        this.pendingTouchEventStart = null;
+        this.pendingTouchEventEnd = null;
+    }
+
+    public setupEvents(element: HTMLElement | Document) {
+        if (this.isTouchDevice()) {
+            element.addEventListener("touchstart", this.handleTouchEventStart.bind(this));
+            element.addEventListener("touchend", this.handleTouchEventEnd.bind(this));
+        } else {
+            element.addEventListener("mousedown", this.handleMouseEventStart.bind(this));
+            element.addEventListener("mouseup", this.handleMouseEventEnd.bind(this));
+            element.addEventListener("mouseenter", this.handleMouseMove.bind(this));
+            element.addEventListener("mousemove", this.handleMouseMove.bind(this));
+            element.addEventListener("mouseleave", this.handleMouseLeave.bind(this));
+        }
+    }
+
+    private isTouchDevice() {
+        return !!('ontouchstart' in window || navigator.maxTouchPoints);
+    }
+
+    private handleMouseEventStart = (event: MouseEvent): void => {
+        this.pendingMouseEventStart = event;
+    }
+
+    private handleTouchEventStart = (event: TouchEvent): void => {
+        if (event.changedTouches.length !== 1) {
+            // This is a multi-touch event. Ignore.
+            return;
+        }
+
+        this.pendingTouchEventStart = event;
+    }
+
+    private handleMouseEventEnd = (event: MouseEvent): void => {
+        if (!this.pendingMouseEventStart) {
+            // Somehow we got an end event without a start event. Ignore it.
+            return;
+        }
+
+        const width = window.innerWidth;
+        const height = window.innerHeight;
+
+        const relativeXDistance = (this.pendingMouseEventStart.clientX - event.clientX) / width;
+        const relativeYDistance = (this.pendingMouseEventStart.clientY - event.clientY) / height;
+
+        // Is the end event in the same place as the start event?
+        if (Math.abs(relativeXDistance) < 0.1 && Math.abs(relativeYDistance) < 0.1) {
+            if (this.pendingMouseEventEnd) {
+                // This was a double click. Let the browser handle it.
+                this.pendingMouseEventStart = null;
+                this.pendingMouseEventEnd = null;
+                return;
+            }
+
+            // This was a single click.
+            this.pendingMouseEventStart = null;
+            this.pendingMouseEventEnd = event;
+            setTimeout(this.handleClick, 200);
```

Comment:
Yes, this is for double-click detection. I initially tried 500ms but it felt too slow to me, particularly for turning pages :(